stl/inc/format

```diff
@@ -4354,6 +4356,268 @@ public:
 private:
     _Fill_align_and_width_specs<_CharT> _Specs;
 };
+
+template <class _CharT>
+struct _Range_specs : _Fill_align_and_width_specs<_CharT> {
+    bool _No_brackets         = false;
+    char _Type                = '\0';
+    bool _Has_underlying_spec = false;
+};
+
+template <class _CharT>
+class _Range_specs_setter : public _Fill_align_and_width_specs_setter<_CharT> {
+public:
+    constexpr explicit _Range_specs_setter(
+        _Range_specs<_CharT>& _Specs_, basic_format_parse_context<_CharT>& _Parse_ctx_)
+        : _Fill_align_and_width_specs_setter<_CharT>(_Specs_, _Parse_ctx_) {}
+
+    constexpr void _On_fill(const basic_string_view<_CharT> _Sv) {
+        if (_Sv == _STATICALLY_WIDEN(_CharT, ":")) {
+            _Throw_format_error("Invalid fill ':' for ranges.");
+        }
+
+        _Fill_align_and_width_specs_setter<_CharT>::_On_fill(_Sv);
+    }
+
+    constexpr void _On_no_brackets() {
+        static_cast<_Range_specs<_CharT>&>(this->_Specs)._No_brackets = true;
+    }
+
+    constexpr void _On_type(const _CharT _Type) {
+        _STL_INTERNAL_CHECK(_Type == 'm' || _Type == 's' || _Type == '?');
+        static_cast<_Range_specs<_CharT>&>(this->_Specs)._Type = static_cast<char>(_Type);
+    }
+
+    constexpr void _On_underlying_spec() {
+        static_cast<_Range_specs<_CharT>&>(this->_Specs)._Has_underlying_spec = true;
+    }
+};
+
+template <class _CharT, class _CallbacksType>
+_NODISCARD constexpr const _CharT* _Parse_range_specs(
+    const _CharT* _Begin, const _CharT* _End, _CallbacksType&& _Callbacks) {
+    if (_Begin == _End || *_Begin == '}') {
+        return _Begin;
+    }
+
+    _Begin = _Parse_align(_Begin, _End, _Callbacks);
+    if (_Begin == _End) {
+        return _Begin;
+    }
+
+    _Begin = _Parse_width(_Begin, _End, _Callbacks);
+    if (_Begin == _End) {
+        return _Begin;
+    }
+
+    if (*_Begin == 'n') {
+        _Callbacks._On_no_brackets();
+        if (++_Begin == _End) {
+            return _Begin;
+        }
+    }
+
+    switch (_CharT _Maybe_type = *_Begin) {
+    case '?':
+        if (++_Begin == _End || *_Begin != 's') {
+            _Throw_format_error("Invalid range-type '?'; did you mean '?s'?");
+        }
+        [[fallthrough]];
+    case 'm':
+    case 's':
+        _Callbacks._On_type(_Maybe_type);
+        if (++_Begin == _End) {
+            return _Begin;
+        }
+    }
+
+    if (*_Begin == ':') {
+        _Callbacks._On_underlying_spec();
+    }
+
+    return _Begin;
+}
+
+template <class _Ty>
+concept _Pair_like_with_formatter =
+    _Is_specialization_v<_Ty, pair> || (_Is_specialization_v<_Ty, tuple> && tuple_size_v<_Ty> == 2);

```

Comment:
I think we should either reuse `_Is_two_tuple` introduced in #4116 or replace it with `_Pair_like_with_formatter`.
https://github.com/microsoft/STL/blob/8dc4faadafb52e3e0a627e046b41258032d9bc6a/stl/inc/format#L2469-L2476